pkg/tickers/time.go

```diff
@@ -0,0 +1,80 @@
+package time
+
+import (
+	"context"
+	"log"
+	"time"
+
+	ocr2keepers "github.com/smartcontractkit/ocr2keepers/pkg"
+)
+
+var (
+	logPrintf = log.Printf
+)
+
+type observer interface {
+	Process(context.Context, Tick) error
+}
+
+// Ticker is a process that runs interval ticks.
+type Ticker interface {
+	Start() error
+	Stop() error
+}
+
+// Tick is the container for the individual tick
+type Tick interface {
+	// GetUpkeeps provides upkeeps scoped to the tick
+	GetUpkeeps(ctx context.Context) ([]ocr2keepers.UpkeepPayload, error)
+}
+
+type getUpkeepsFn func(context.Context, time.Time) (Tick, error)
+
+type timeTicker struct {
+	interval time.Duration
+	ticker   *time.Ticker
+	observer observer
+	getterFn getUpkeepsFn
+	ctx      context.Context
+	cancelFn context.CancelFunc
+}
+
+func NewTimeTicker(interval time.Duration, observer observer, getterFn getUpkeepsFn) *timeTicker {
+	ctx, cancelFn := context.WithCancel(context.Background())
+	t := &timeTicker{
+		interval: interval,
+		ticker:   time.NewTicker(interval),
+		observer: observer,
+		getterFn: getterFn,
+		ctx:      ctx,
+		cancelFn: cancelFn,
+	}
+	go t.Start()
+	return t
+}
+
+func (t *timeTicker) Start() {
+	for {
+		select {
+		case tm := <-t.ticker.C:
+			func() {
+				ctx, cancelFn := context.WithTimeout(t.ctx, t.interval)
+				defer cancelFn()
```

Comment:
so this cancelFn is different with the cancelFn in timeTicker as in L50, what are the use cases for both?

Reply:
In go, when you create a context, you must specify a parent context from which the new context will be derived. So the cancelFn on line 50 is the parent context, and then on line 61, we create a new child context from the parent context. The child context has a timeout equal to the tick interval, to ensure that the subsequent operations complete before the ticker's next tick. When we call Stop() on this ticker, we cancel the parent context, which in turn cancels the child context.